sequila/sequila-core/src/physical_planner/joins/interval_join.rs

```diff
@@ -1076,28 +1120,38 @@ impl IntervalJoinStream {
         let start = evaluate_as_i32(self.right_interval.start(), &state.batch)?;
         let end = evaluate_as_i32(self.right_interval.end(), &state.batch)?;
 
-        let mut builder_left = PrimitiveBuilder::<UInt32Type>::new();
-
-        let mut rle_right: Vec<u32> = Vec::with_capacity(self.hashes_buffer.len());
-        let mut pos_vect: Vec<u32> = Vec::with_capacity(100);
-        for (i, hash_val) in self.hashes_buffer.iter().enumerate() {
-            build_side
-                .hash_map
-                .get(*hash_val, start.value(i), end.value(i), |pos| {
-                    pos_vect.push(pos as u32);
-                });
-            rle_right.push(pos_vect.len() as u32);
-            builder_left.append_slice(&pos_vect);
-            pos_vect.clear();
-        }
-        let left_indexes = builder_left.finish();
-        let mut index_right = Vec::with_capacity(left_indexes.len());
-        for i in 0..rle_right.len() {
-            for _ in 0..rle_right[i] {
-                index_right.push(i as u32);
+        let mut left_indexes: Vec<u32> = Vec::with_capacity(self.hashes_buffer.len());
+        let mut right_indexes: Vec<u32> = Vec::with_capacity(self.hashes_buffer.len());
+
+        match &build_side.hash_map {
+            IntervalJoinAlgorithm::Coitrees {
+                hashmap: _,
+                use_sorted: true,
+            } => {
+                build_side.hash_map.process(
+                    &self.hashes_buffer,
+                    &start,
+                    &end,
+                    |(right_pos, left_pos)| {
+                        left_indexes.push(right_pos as u32);
+                        right_indexes.push(left_pos as u32);
+                    },
+                );
+            }
+            _ => {
+                for (i, hash_val) in self.hashes_buffer.iter().enumerate() {
+                    build_side
+                        .hash_map
+                        .get(*hash_val, start.value(i), end.value(i), |pos| {
+                            left_indexes.push(pos as u32);
+                            right_indexes.push(i as u32);
+                        })
+                }
             }
         }
-        let right_indexes = PrimitiveArray::from(index_right);
+
+        let left_indexes: UInt32Array = left_indexes.into();
```

Comment:
@mwiewior I refactored it for now to make it clear for me what is going on there. i'll revert it later.